File: StriverDP/Triangle.cpp
/*
https://www.codingninjas.com/codestudio/problems/triangle_1229398?source=youtube&campaign=striver_dp_videos&leftPanelTab=0
*/
#include <bits/stdc++.h> 
int minimumPathSum(vector<vector<int>>& triangle, int n){
	vector<vector<int>> dp(n,vector<int>(n,0));
    dp[n-1] = triangle[n-1];
    for(int i = n-2 ; i >= 0 ; i--){
        for(int j = i ; j>=0 ; j--){
            int d = triangle[i][j] + dp[i+1][j];
            int diag = triangle[i][j] + dp[i+1][j+1];
            dp[i][j] = min(d,diag);
        }
    }
    return dp[0][0];
}

File: StriverDP/triangle.py
'''
https://www.codingninjas.com/codestudio/problems/triangle_1229398?source=youtube&campaign=striver_dp_videos&leftPanelTab=0
'''

from os import *
from sys import *
from collections import *
from math import *

def minimumPathSum(triangle, n):
    front = triangle[-1]
    for i in reversed(range(n-1)):
        curr = [0 for k in range(i+1)]
        for j in reversed(range(i+1)):
            up = front[j]
            diag = front[j+1]
            curr[j] = triangle[i][j] + min(up, diag)
        front = curr.copy()
    return front[0]
